### PR TITLE
Ch7 point to generic derive guide and include manual solutions too

### LINKED ISSUE #232: Ch 7 - Review tree show instance and solutions
The exercise is:

> 1. (Easy) Write Eq and Show instances for the following binary tree data structure:
>
> ```purs
> data Tree a = Leaf | Branch (Tree a) a (Tree a)
> ```

The solution is implemented using generics. By this point `derive` isn't discussed not to mention generics. There's discussion to include at least `derive` in Ch6 #149, but not generics. Thus the solutions should be "manual" solutions.

Also the implementation of the show instance is pretty subjective so the reader needs a bit more details on how the solution should look like otherwise he's forced to look at the test suite. Including an output example would do the trick, something like: `(Branch (Branch Leaf 5 Leaf) 4 (Branch Leaf 6 Leaf))` or something like that maybe.

### PR DESCRIPTION
**Pull request by milesfrain**
Fixes #232 

I'd like to promote the derived solution by letting the manual solution be the commented-out one for reference.

### CODE REVIEW

File: exercises/chapter7/test/no-peeking/Solutions.purs [hunk 1]
```diff
@@ -84,6 +85,29 @@ instance eqTree :: Eq a => Eq (Tree a) where
...
 
+{-
+-- Solution using manually-written instances:
+
+instance eqTree :: Eq a => Eq (Tree a) where
+  eq Leaf Leaf = true
+  eq (Branch t1a va t2a) (Branch t1b vb t2b) =
+    t1a == t1b
+      && va == vb
+      && t2a == t2b
```
Comment (milesfrain) [inline exercises/chapter7/test/no-peeking/Solutions.purs:96]: Another autoformatting struggle.

I ideally want something like:
```purs
  eq (Branch t1a va t2a) (Branch t1b vb t2b)
      =  t1a == t1b
      && va  == vb
      && t2a == t2b
```

These are the only purty-friendly options I could find:
```purs
  eq (Branch t1a va t2a) (Branch t1b vb t2b) =
    t1a == t1b
      && va
      == vb
      && t2a
      == t2
```
```purs
  eq (Branch t1a va t2a) (Branch t1b vb t2b) = t1a == t1b && va == vb && t2a == t2b
```
```purs
  -- Unnecessary parens
  eq (Branch t1a va t2a) (Branch t1b vb t2b) =
    (t1a == t1b)
      && (va == vb)
      && (t2a == t2b)
```

**Comment by razcore-art** [comment]
Morning! :)

1. Just to make sure a couple of things. Are you planning on adding a discussion on deriving before this exercise as well or is this all?

I think there needs to be a few words about `derive` in general and `derive newtype` before getting to generic deriving including a link to that documentation that doesn't reference generic deriving specifically.

2. The compiler can derive `Eq` without generics: `derive instance eqTree :: Eq a => Eq (Tree a)` so one wonders why you'd choose one over the other. I know generics by its nature of being generic has to do more work so it's slower than the other options.

So there's a couple of things to think about.

**Comment by milesfrain** [comment]
All good points. And definitely missed-out on non-generic `Eq`. I can't think of any reasons for using generic `Eq`.

Ch6 now seems like a more appropriate place for mentioning all the deriving options. Thinking of adding a note (and link to guide) on basic/newtype/generic deriving immediately before this [Ch6 exercise section](https://book.purescript.org/chapter6.html#exercises-1). Then creating a few more exercises to cover writing instances involving:
- ADTs, Records, newtypes
- manual and derived (basic/newtype/generic) instances
- A variety of classes: `Eq`, `Ord`, `Show`, `Foldable`, etc.

The existing `Complex` number exercise covers a few of those topics, but we can extend it further to include:
- manual `Show`
  - Expect `2.7+3.4i` formatting
- basic derive for `Eq`
- manual `Semiring`
  - requires custom `mul`, but could have otherwise gotten away with newtype for `add`, `zero`, and `one`.
- newtype derive `Ring`

Might be good to note that wrapping records with `newtype` is not always necessary (discussed in #118). For example, `type Point = {x :: Number, y :: Number}` can be used for all the above purposes as-is (assuming default `Show` is okay). 

For ADT + generic coverage, `Show` seems like the prime use case (until generic JSON in later chapters). Some options for that:
- Simplified `data IntTree = Leaf | Branch IntTree Int IntTree` from [Ch7](https://book.purescript.org/chapter7.html#exercises-2).
- `Shape` from Ch5.
  - We'd need to replace the [first exercise](https://book.purescript.org/chapter6.html#exercises) with something else, which I'm in favor of for breaking the dependency on the previous chapter. Could be a good opportunity for newtype-wrapped `Point` with `Field` support (assuming division makes sense) added in the next section.
- Simplified `data ExtendedInt = Finite Int | Infinite` cloned from the [next exercise section](https://book.purescript.org/chapter6.html#exercises).
  - Could also ask readers to write `Eq`, `Ord`, and `Semiring` instances in addition to `Show`.

 Let me know if you have any suggestions. Exercise gaps for this section are still `Semigroup` `Monoid`, `Foldable`, and `Functor`.

**Comment by razcore-art** [comment]
This looks like solid points to me. Ch6 is a good place for derive discussion & exercises. I don't think I have anything to add right now, but if I think of anything i'll let you know.

**Comment by milesfrain** [comment]
Let's revisit this once #240 is merged

**Pull request merged**